## import dependencies

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## load data

In [20]:
df = pd.read_csv('../data/raw.csv')
df.head()

ID_Pengiriman              Tanggal  Waktu_Keberangkatan  \
0     DEL-00001  2023-10-09 10:28:00  2023-10-09 10:28:00   
1     DEL-00002  2023-09-15 17:07:00  2023-09-15 17:07:00   
2     DEL-00003  2023-09-21 13:57:00  2023-09-21 13:57:00   
3     DEL-00004  2023-09-19 13:10:00  2023-09-19 13:10:00   
4     DEL-00005  2023-09-11 14:52:00  2023-09-11 14:52:00   

             Waktu_Kedatangan  Lat_Jemput  Lon_Jemput  Lat_Tujuan  Lon_Tujuan  \
0  2023-10-09 12:09:35.444386   -6.187770  106.884394   -6.324869  106.783430   
1  2023-09-15 17:53:45.354053   -6.140012  106.898952   -6.140302  106.711119   
2  2023-09-21 14:05:42.112794   -6.101567  106.948339   -6.204143  106.927419   
3  2023-09-19 13:53:23.355129   -6.112538  106.778596   -6.292112  106.655991   
4  2023-09-11 15:15:47.785700   -6.085218  106.912228   -6.172220  106.816521   

     Jarak_KM_Terukur  ID_Kurir Tipe_Kendaraan Pengalaman_Kurir  Jumlah_Paket  \
0   39.44772552617686        21          Motor         Menengah            37   
1  19.092557707822586         7          Motor         Menengah             5   
2                 1.0         1          Motor         Menengah             5   
3  23.393817976950878         1            MTR           Rendah            11   
4  17.144595029161525        18      Mobil_Van           Rendah            16   

  Kondisi_Cuaca Kondisi_Lalu_Lintas Tipe_Area_Tujuan  Durasi_Pengiriman_Menit  \
0         Badai              Sedang        Perumahan               101.590740   
1  Hujan_Ringan              Ringan        Perumahan                46.755901   
2         Cerah              Sedang        Perumahan                 8.701880   
3         Cerah              Sedang        Perumahan                43.389252   
4         Cerah              Ringan        Perumahan                23.796428   

   ETA_Awal_Menit  Status_Keterlambatan  
0       63.494212                     1  
1       42.690170                     0  
2        7.614145                     0  
3       37.965596                     0  
4       24.986250                     0

## Cleaning Data

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 19 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID_Pengiriman            10000 non-null  object 
 1   Tanggal                  10000 non-null  object 
 2   Waktu_Keberangkatan      10000 non-null  object 
 3   Waktu_Kedatangan         10000 non-null  object 
 4   Lat_Jemput               10000 non-null  float64
 5   Lon_Jemput               10000 non-null  float64
 6   Lat_Tujuan               9800 non-null   float64
 7   Lon_Tujuan               10000 non-null  float64
 8   Jarak_KM_Terukur         10000 non-null  object 
 9   ID_Kurir                 10000 non-null  int64  
 10  Tipe_Kendaraan           10000 non-null  object 
 11  Pengalaman_Kurir         10000 non-null  object 
 12  Jumlah_Paket             10000 non-null  int64  
 13  Kondisi_Cuaca            10000 non-null  object 
 14  Kondisi_Lalu_Lintas    

1. tipe data yang salah
    - tanggal
    - waktu keberangkatan
    - waktu kedatangan
    - jarak km terukur
2. nilai kosong
    - lon tujuan
    - durasi pengiriman paket

In [29]:
df.describe()

Lat_Jemput    Lon_Jemput   Lat_Tujuan    Lon_Tujuan      ID_Kurir  \
count  10000.000000  10000.000000  9800.000000  10000.000000  10000.000000   
mean      -6.197676    106.798379    -6.200301    106.800355     24.933600   
std        0.086216      0.086299     0.086885      0.086432     14.105293   
min       -6.349986    106.650002    -6.349995    106.650003      1.000000   
25%       -6.271467    106.722851    -6.275431    106.725907     13.000000   
50%       -6.196705    106.798648    -6.200916    106.801431     25.000000   
75%       -6.124507    106.872333    -6.124130    106.874930     37.000000   
max       -6.050030    106.949937    -6.050008    106.949987     49.000000   

       Jumlah_Paket  Durasi_Pengiriman_Menit  ETA_Awal_Menit  \
count  10000.000000              9900.000000    10000.000000   
mean      21.858200                47.440540       40.337984   
std       10.057107                62.566147       24.940046   
min        5.000000                 5.000000      -21.089756   
25%       13.000000                22.292487       20.602444   
50%       22.000000                43.100462       40.027153   
75%       31.000000                64.924636       59.093225   
max       39.000000              1967.535417      119.756683   

       Status_Keterlambatan  
count          10000.000000  
mean               0.046500  
std                0.210576  
min                0.000000  
25%                0.000000  
50%                0.000000  
75%                0.000000  
max                1.000000

1 hari ada 1440 menit, tapi nilai maksimal durasi pengiriman menit itu 1967.5 berarti waktu pengiriman itu lebih dari 1 hari

In [22]:
df.isnull().sum()

ID_Pengiriman                0
Tanggal                      0
Waktu_Keberangkatan          0
Waktu_Kedatangan             0
Lat_Jemput                   0
Lon_Jemput                   0
Lat_Tujuan                 200
Lon_Tujuan                   0
Jarak_KM_Terukur             0
ID_Kurir                     0
Tipe_Kendaraan               0
Pengalaman_Kurir             0
Jumlah_Paket                 0
Kondisi_Cuaca                0
Kondisi_Lalu_Lintas          0
Tipe_Area_Tujuan             0
Durasi_Pengiriman_Menit    100
ETA_Awal_Menit               0
Status_Keterlambatan         0
dtype: int64

In [23]:
df.duplicated().sum()

np.int64(0)

In [26]:
df[df['Lat_Tujuan'].isna()].head()

ID_Pengiriman              Tanggal  Waktu_Keberangkatan  \
127     DEL-00128  2023-10-03 18:58:00  2023-10-03 18:58:00   
190     DEL-00191  2023-10-27 19:13:00  2023-10-27 19:13:00   
239     DEL-00240  2023-10-05 18:24:00  2023-10-05 18:24:00   
255     DEL-00256  2023-10-08 09:07:00  2023-10-08 09:07:00   
256     DEL-00257  2023-10-18 19:00:00  2023-10-18 19:00:00   

               Waktu_Kedatangan  Lat_Jemput  Lon_Jemput  Lat_Tujuan  \
127  2023-10-03 19:24:45.531760   -6.269799  106.704647         NaN   
190  2023-10-27 21:19:25.129401   -6.328868  106.848336         NaN   
239  2023-10-05 19:31:09.173323   -6.173669  106.940136         NaN   
255  2023-10-08 09:51:29.567022   -6.324640  106.669052         NaN   
256  2023-10-18 19:51:05.259561   -6.209182  106.705719         NaN   

     Lon_Tujuan    Jarak_KM_Terukur  ID_Kurir Tipe_Kendaraan Pengalaman_Kurir  \
127  106.904497    8.77306145170068        20            Van         Menengah   
190  106.819132   44.41933018104206        11          Motor           Rendah   
239  106.759825  30.658789063915187        46      Mobil_Van         Menengah   
255  106.755005   28.40339066397479        41          Motor         Menengah   
256  106.753942            26.15 KM        16          Motor           Tinggi   

     Jumlah_Paket Kondisi_Cuaca Kondisi_Lalu_Lintas Tipe_Area_Tujuan  \
127             5         Cerah               Padat        Perumahan   
190             9         Cerah               Padat         Industri   
239            24  Hujan_Ringan              Ringan        Perumahan   
255             6         Cerah              Ringan        Perumahan   
256            30  Hujan_Ringan              Ringan           Bisnis   

     Durasi_Pengiriman_Menit  ETA_Awal_Menit  Status_Keterlambatan  
127                26.758863       17.560504                     0  
190               126.418823       82.962353                     1  
239                67.152889       61.313507                     0  
255                44.492784       46.717423                     0  
256                51.087659       51.828060                     0

In [28]:
df[df['Durasi_Pengiriman_Menit'].isna()].head()

ID_Pengiriman              Tanggal  Waktu_Keberangkatan  \
45      DEL-00046  2023-10-09 16:40:00  2023-10-09 16:40:00   
173     DEL-00174  2023-10-17 09:54:00  2023-10-17 09:54:00   
321     DEL-00322  2023-10-20 18:00:00  2023-10-20 18:00:00   
419     DEL-00420  2023-10-01 07:04:00  2023-10-01 07:04:00   
455     DEL-00456  2023-09-05 19:46:00  2023-09-05 19:46:00   

               Waktu_Kedatangan  Lat_Jemput  Lon_Jemput  Lat_Tujuan  \
45   2023-10-09 17:36:57.852013   -6.146412  106.663926   -6.194724   
173  2023-10-17 09:59:00.000000   -6.068798  106.859938   -6.171705   
321  2023-10-20 18:42:42.356223   -6.191782  106.697672   -6.136776   
419  2023-10-01 07:43:57.019557   -6.243583  106.939672   -6.189150   
455  2023-09-05 20:45:14.236966   -6.115285  106.922201   -6.074242   

     Lon_Tujuan    Jarak_KM_Terukur  ID_Kurir Tipe_Kendaraan Pengalaman_Kurir  \
45   106.854092  15.615452775705362        35          Motor         Menengah   
173  106.808994   7.066344283939976        18      Mobil_Van           Tinggi   
321  106.896711  23.005796025097048        19      Mobil_Van         Menengah   
419  106.809107   13.12296521051206        23          Motor         Menengah   
455  106.686738  33.594953941317165        13          Motor           Rendah   

     Jumlah_Paket Kondisi_Cuaca Kondisi_Lalu_Lintas Tipe_Area_Tujuan  \
45             29         Cerah               Padat         Industri   
173            24         Cerah              Ringan        Perumahan   
321            10         Cerah              Ringan           Bisnis   
419            36         Cerah               Padat        Perumahan   
455            27         Cerah              Sedang        Perumahan   

     Durasi_Pengiriman_Menit  ETA_Awal_Menit  Status_Keterlambatan  
45                       NaN       37.382756                     0  
173                      NaN       -0.402051                     0  
321                      NaN       44.841234                     0  
419                      NaN       26.217401                     0  
455                      NaN       51.832622                     0